### ISSUE TITLE
1.4.13 Content on Hover or Focus - dismissable and persistent clarification

### ISSUE THREAD

**Issue by smhigley** [w3c/wcag#914]
In attempting to make good tooltip examples and talking with developers about tooltips and 1.4.13, I've run across one blocking issue with the dismissable requirement and one suggestion for persistence:

**Dismissable**

The only dismissable example provided in the Understanding section mentions using the escape key (or any other keyboard interaction) to dismiss content on hover or focus. However, this has certain blocking technical limitations, since the element capturing the keyboard event isn't likely to be the same element displaying the popup on pointer hover. It is possible to have an escape event listener on the body tag, but this will fail for tooltips appearing within modals or tooltips sharing a page with any other component that uses escape (e.g. menus, comboboxes, etc.) and may have keyboard focus. At worst, a user attempting to dismiss a tooltip within a modal will not be able to do so without dismissing the entire modal.

The only practical solution I've found has been to include a close button on the tooltip, but this violates the requirement to dismiss without pointer movement (related to  #409). It's possible to prevent the tooltip from re-appearing on hover after actively closing it, which seems to meet the spirit of the criterion, but still violates the text.

Would it be possible to either modify the pointer dimissable requirement to allow that technique, or provide an alternate example technique that works for pointer users?

**Persistent**

This came up when evaluating how 1.4.13 applies to desktop tooltips through EN 301 549. Windows has had (very long) timeouts built into desktop tooltips for a good while, and adding a Settings option to adjust or remove tooltip timeouts was suggested for this criterion. This still technically violates 1.4.13, though it seems internally consistent with 2.2.1/Timing Adjustable. Would there be any interest in modifying the Persistent requirement to either refer to Timing Adjustable or allow a user-adjustable timeout?

**Comment by JAWS-test** [comment]
> Would it be possible to either modify the pointer dismissable requirement to allow that technique, or provide an alternate example technique that works for pointer users?

First I would like to know if it is possible to design [SCR39](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR39) so that the page contains multiple tooltips and other elements that are closed with ESC, and tooltips within pop-ups - and still work.

**Comment by alastc** [comment]
If `esc` is generally caught and dismisses 'open' things, that is likely to be a good thing except in the case of nested items.

Having a dismissable item nested within another dismissable item probably isn't a good idea!

Another avenue is to use the exception within dismissable: "unless the additional content ... does not obscure or replace other content". I.e. if the tooltip is reliably in an area that does not cover other content, then it does not need to be dismissable.

For the persistent point, it is applying to all sorts of pop-overs, and in the case of large boxes of content I don't think many of the LV users would want it to time out. Browser-native tooltips and explicitely excepted, so it is only small custom tooltips that fall into this area.

We can ask, but I suspect the LV folk would rather hover wasn't used for this sort of functionality, it is difficult to use when magnified. Given the popularity of touch-screen usage, it seemed likely that click/tap would be a better approach all-around.

**Comment by JAWS-test** [comment]
> Having a dismissable item nested within another dismissable item probably isn't a good idea!

I wonder if it is possible to close with ESC only the element that was shown by hover depending on the keyboard focus and current mouse position - or to close with ESC only the pop-up if no nested elements are shown by focus or hover. If a content is displayed via hover or keyboard focus, ESC would always close this element first and the pop-up would be closed at the second ESC.

**Comment by smhigley** [comment]
@alastc I think perhaps I'm perhaps reading too much into your recommendation -- my first read of this:

>If esc is generally caught and dismisses 'open' things, that is likely to be a good thing except in the case of nested items.
> 
> Having a dismissable item nested within another dismissable item probably isn't a good idea!

is that it's an acceptable side effect of 1.4.13 be to forbid any content on hover or focus within a modal. While I personally dislike tooltips and other hover/focus popups and would be OK with forbidding them outright, it would be quite a hard sell to tell all product teams that they can never put a tooltip inside a modal. Is there another way to meet 1.4.13 that allows tooltips in modals without relying on not overlapping other content?

From a practical perspective, I've found the "does not obscure or replace other content" exception almost never works, since at some screen size between desktop and mobile it's likely that a tooltip, even a short one, will overlap something as the page reflows. Especially when designing reusable tooltips, it's impossible to know whether one will be used either within a modal or in a way that overlaps other content (particularly at small screen sizes).

To the "it's ok to close all open things" point -- I don't disagree with the theoretical user experience, and it could be technically feasible if the user's focus is on an element without a popup. I run into some issues in specific cases that still seem blocking, since the web author doesn't have control over where the user's focus may be at any given point.

Since any component that captures and uses `escape` must stop propagation of escape key presses for the same reason I raised the issue -- a combobox won't know if it's on a static page or within a modal, so it needs to always `.stopPropagation` of escape key events to prevent unintended side effects. So if a user is in a situation where they've clicked into a combobox (or any other component with escape behavior) and then go explore a tooltip somewhere else, escape would not be able to close the tooltip; it would be consumed by the combo and would never reach the `body` tag. My current impression is that this means a hover-triggered popup will not be able to rely on `escape` not only within modals, but also when sharing a page with components like comboboxes, menus, etc. that may use escape. Though please let me know if I'm missing anything that would enable a generic dismissable tooltip (or other popup) on hover, or if the intent is really as prohibitive as it seems (to be fair, I'd love to get rid of all hover content, full stop 😄).

I definitely get the reluctance for timing adjustable pop-overs, it does seem like a benefit in only a small subset of cases, and definitely more feasible to have user-adjustable timing settings in an OS than in a web app. The browser-native tooltip is actually an interesting case study for this though -- I know web content authors are exempt when it comes to `title` attributes, but are browser vendors on the hook? That is, are all the major browsers currently in violation of 1.4.13 (or C.11.1.4.13) through EN 301 549?

**Comment by JAWS-test** [comment]
I have created an [example](https://codepen.io/jaws-test/pen/MWWwmEp) where the first ESC closes all tooltips and the second ESC closes the pop-up. If no tooltips are open, the first ESC closes the pop-up. The example shows that it is not a problem to evaluate ESC contextually.

**Comment by JAWS-test** [comment]
In the meantime, I also think that SC is difficult to implement in practice:
* Hover over an element in an iFrame: ESC key doesn't work if keyboard focus is not in the iFrame
* Hover over an element: ESC key does not work if focus is not on the page but on an element of the browser (or even in another application)
* Hover over an element and keyboard focus on another element: which of the two displayed contents should ESC close?

**Comment by smhigley** [comment]
Is there any update on this? This requirement is still causing a non-trivial amount of effort put into both investigating whether it is achievable from a technical perspective in various contexts, and what the options are if and when it is not.

Specifically, these are the cases we have found where dismissing hovered pointers via escape is not viable:
- content within iframes
- content within modals, where the modal is a component from a UI framework
- content that shares a page with a component that traps keyboard events, such as a toolbar with tooltips above an in-browser code editor
- developing UI frameworks that build reusable components that are not aware of their context (in the case of components with tooltips), or their content (in the case of modals)
- any product that uses modals, slidepanes, or tooltips from a UI framework

I also conducted a small-scale usability study with 12 participants, and found that well-placed close buttons were preferable to the escape or control keys for dismissing tooltips by low vision users. Most participants were using magnification at around 800%.

Right now, my personal recommendation is that teams should not invest time into attempting to follow the pointer dismissal specifics, but of course that causes issues when WCAG is a legal requirement. Is there any chance this will be fixed in 2.2, or get an updated Understanding doc with more viable possibilities?

For reference, this is the tooltip close button design that performed better than keyboard dismissal in the usability study:
![icon button tooltip with a close button offset from the tooltip on the closest side to the trigger button](https://user-images.githubusercontent.com/3819570/73972948-87fe1b80-48d6-11ea-8727-acd398cc25ba.png)

**Comment by alastc** [comment]
Hi @smhigley,

Thank you, that's very well conveyed. (Arg, I read that and it sounds sarcastic, it is sincere, I just can't think of another way to say: awesome bug report, it is really clear!) 

I'm going to ask the Low Vision Task Force for comment (and @steverep if he is around).

One thing that jumps out to me is that “without moving pointer hover or keyboard focus” was intended to mean: move **away from the target**. 

I think that might have been ‘refined out’ at some stage for clarity. Or perhaps it was that a close button within the target would be very complex, as you’d have to trigger on entry to the target only.

With that meaning, the solution where the ‘close’ button overlaps with the target would work I think?

**Comment by RobGallo** [comment]
One thing that needs to happen with this entire discussion is the complete abandonment of the escape key as a tooltip dismissal mechanism. It's folly.

A blind or low vision user cannot be expected to know at all times if a tooltip is present or not. Thus the behavior of the escape key will, in practice, be unpredictable. Since the escape key is often a key with consequences, it is not an acceptable user experience to link a possibly unknowable behavior to such a key.

**Comment by alastc** [comment]
> A blind or low vision user cannot be expected to know at all times if a tooltip is present or not.

The requirement that there is a mechanism to close pop-over content came from low-vision users. (If someone is completely blind there are generally different problems with tooltips.)

`Esc` key is one method, and tooltips are a sub-set of the possible pop-over content. There are other methods, and there are other types of content which might suit different methods.

Take the tooltips above this edit box (aA, B, i) for example, they have tooltips that reliably do not cover other content, so no need for another mechanism.

I do talk the point that the UX for low vision users can be tricky, but it is something asked for because the UX of pop-over content is worse.

We took mention of the close button out because it seemed to difficult to do, but if @smhigley created and tested it, perhaps that is a better solution for that type of content.

**Comment by RobGallo** [comment]
My point is that the escape key is not a useable method to dismiss any non-modal or non-focused content. WCAG needs to be extremely clear about that. And regardless if LV users asked for the criterion, the negative impact of using escape on blind users will be the same.

**Comment by alastc** [comment]
A typical instance we get in audits is a mega-menu, and in that case using esc is a reasonable method. It isn't very discoverable, but if you need it, it is there.

It doesn't generally affect screenreaders because it either isn't triggered, or is triggered by focus and esc should put your back on the opening element.

**Comment by RobGallo** [comment]
In the case you describe, the menu sounds like


  1.  It has focus
  2.  It is intended that a user interact with it

Neither of which is true of a non-modal, non-focusable overlay.

In the case where the user is interacting with an element with focus directly, escape is an acceptable solution. But those situations are not the ones we’re talking about. Escape is never okay for any content under the tooltip umbrella: non-modal, non-focusable popups or overlays.

If you’re saying that the menu is triggered by hover, then I’ll say, and rightly so, that a mouse hover should never trigger a focus change, or the need for a focus change. That is an unexpected change of focus, which is disorienting for screen reader users.

**Comment by smhigley** [comment]
@alastc thanks for the response!  Do you think updated wording to allow for overlaid close buttons could make it into 2.2, at least if it's not controversial?

One other piece of clarification: would the requirement be that the close button overlays the hit target, even if the boundaries of the hit target aren't visible? E.g. in the case of links or borderless buttons, would this still pass?

I can foresee designers griping a bit about the specificity of requiring the close button to overlap the target vs. some measure of proximity or ergonomics, but this does at least give a path forward, so I'll take it 😁.

Thanks again!

**Comment by alastc** [comment]
@smhigley I'm not sure if it would be better to update 2.1 (errata) or adjust in 2.2, we'd have backwards incompatibility issues with just adjusting in 2.2. We'll discuss that.

Also, where I said "move away from the target" above, I had meant the trigger, i.e. the element that triggers the pop-over content. (I was thinking in terms of mouse targets, sorry.)

The overlapping example fulfils that either way, but just to be clear about where the requirement came from.

**Comment by smhigley** [comment]
@alastc I had meant the element that triggers the pop-over as well 😄. To add some visuals, would the tooltip in the first screenshot below pass?:
![tooltip positioned below a link with no visible border, with the close button visually below the tooltip text](https://user-images.githubusercontent.com/3819570/74194966-23f99100-4c0f-11ea-94d5-29b916dbb459.png)

This second screenshot shows the same link/tooltip, but with a background color to demonstrate that the close button does intersect with the boundaries of the link element, even if it is not visually apparent in the first screenshot:
![same tooltip image, but with a colored link background showing the close button overlapping the link](https://user-images.githubusercontent.com/3819570/74195031-42f82300-4c0f-11ea-9d1a-00d900356fe3.png)

**Comment by joe-watkins** [comment]
Since sub-menus are clearly called out in the understanding document:

> A typical instance we get in audits is a mega-menu, and in that case using esc is a reasonable method. It isn't very discoverable, but if you need it, it is there.

1. Can anyone provide a single accessible example of this recommended approach in the wild where the mega menu / or site navigation / sub-menu, which has hover as one of the events to show/hide sub-menu items, can be closed with the escape key?

2. Does this ARIA menu fail 1.4.13 because sub-menu items can't be dismissed with escape key as it obscures content beneath it?
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html

3. Can anyone provide an alternative mechanism for closing sub-menu items that isn't the escape key and that doesn't require moving the mouse?

@smhigley It's my interpretation that if the user has to move the mouse to dismiss something it would be a failure. So having to move the mouse to click on a close control would fail. 

> Failure to make content dismissable without moving pointer hover or keyboard focus

**Comment by idoros** [comment]
>  close button does intersect with the boundaries of the link element

Notice that having such intersection means that in some cases after clicking the close button, the mouse will be over the tooltip anchor again and might trigger it immediately or on move. Tooltip delay helps, but still, even moving the mouse for a bit and then keeping it in the boundaries will definitely pop the tooltip again.

**Comment by innoticFR** [comment]
A question relevant to the previous example.

The default behavior in Google Chrome/Windows for the `title` attribute of the `<abbr>` tag is the following:

- Hover shows a tooltip.
- This can be dismissed using the Esc key.
- This can also be dismissed by clicking on the element (without moving the pointer).

Does a click-to-dismiss mechanism satisfy WCAG Success Criterion 1.4.13? Or could it lead to other failures like 3.2.1?

**Comment by detlevhfischer** [comment]
> The default behavior in Google Chrome/Windows for the title attribute of the <abbr> tag is the following:

I think `title` would fall under the user agent exception of 1.4.13. So no failure here even if user agents may differ in the way they are exposing or removing / hiding `title`.

**Comment by innoticFR** [comment]
> > The default behavior in Google Chrome/Windows for the title attribute of the  tag is the following:
> 
> I think `title` would fall under the user agent exception of 1.4.13. So no failure here even if user agents may differ in the way they are exposing or removing / hiding `title`.

But does replicating such common behavior (click to dismiss, without a close button)  with a custom component (dropdown menu, tooltip) is sufficient?

**Comment by detlevhfischer** [comment]
> But does replicating such common behavior (click to dismiss, without a close button) with a custom component (dropdown menu, tooltip) is sufficient?

 While ESC is the most common way of meeting the close-part of 1.4.13, I guess a click without moving focus would meet that part of the requirement in 1.4.13 - you would in addition need to be able to move the pointer over the content without it disappearing. You may fail to meet other requirements though, for example, if there is no keyboard way of closing the popup.

**Comment by innoticFR** [comment]
> You may fail to meet other requirements though, for example, if there is no keyboard way of closing the popup.

Good question! Is it mandatory to have a keyboard way to dismiss a content appearing only on hover? This would mean having to click first on the element to set the keyboard focus there, which would already hide the popup element which has the onclick handler. Or to wait for the `Escape` event listerner on the body element.

Otherwise, if the element can also be automatically opened on focus, this means the element is focusable and a simple toggleable element, activated with the onclick handler (enter, spacebar, click) could toggle the visibility of the controlled element, which would answer the previous point.

I can imagine a scenario where a user is using the mousewheel to scroll, which opens a popup, and is unable to click using his right hand to close it. But there's also a scenario where the user is only using a mouse or can't use their two hands. Requiring to press the Escap key is already challenging.

**Comment by GreggVan** [comment]
@innoticFR   Good question! Is it mandatory to have a keyboard way to dismiss a content appearing only on hover? 

Not mouse hover - no.   The requirements are that it all be accessible via keyboard.  so any content hidden by KEYBOARD FOCUS  should be dismiss-able from keyboard.   But not a popup from a pointer.  

Although there may be edge conditions where it is useful to someone -- it is not a requirement of 'all functionality from keyboard only'   note the only.

**Comment by scottaohara** [comment]
the normative text doesn't overtly state that - and i think you're referring to note 3 in your response - but that also seems out of scope to the question asked.

The understanding doc seems to imply that a keyboard means of dismissing would be expected for hover as well:

>When magnified, the portion of the page visible in the viewport can be significantly reduced. Mouse users frequently move the pointer to pan the magnified viewport and display another portion of the screen. However, almost the entire portion of the page visible in this restricted viewport may trigger the additional content, making it difficult for a user to pan without re-triggering the content. A keyboard means of dismissing the additional content provides a workaround.

after that paragraph the understanding doc then starts talking about people navigating primarily by keyboard.... so, per the normative text and the information i cited, it seems the intent would be that it would be necessary that a keyboard mechanism be available even if the popup only appears via hover.  

for content that appears when it is focused (re: note 3 and referencing a skip-link) that content it out of scope for the SC

**Comment by patrickhlauke** [comment]
> Not mouse hover - no. The requirements are that it all be accessible via keyboard. so any content hidden by KEYBOARD FOCUS should be dismiss-able from keyboard. But not a popup from a pointer.

the SC requires that if something appeared on pointer hover, there must be a way to dismiss it that doesn't involve moving the pointer away.

so generally, the most accepted way to achieve that is to rely on keyboard `Esc` or similar. technically yes, clicking would also be a valid way to achieve that.

> The requirements are that it all be accessible via keyboard

not directly, no. 1.4.13 says nothing of the sort. this would be a 2.1.1 concern.

**Comment by GreggVan** [comment]
agree to all your points @patrickhlauke 

I  was not referring to ONLY 1.4.13  but to WCAG2 as a whole

**Comment by innoticFR** [comment]
Thanks @patrickhlauke 

If I understand correctly:

1. out of the scope of 1.4.13, SC 2.1.1 says that everything should be accessible from keyboard only. This does not mandate that something activated using another peripheric (mouse pointer in this case) could be dismissed using a keyboard. 
2. A toggle button activating a dropdown menu could also activate this submenu on mouseover, and discard it with a click on the same toggle button.
3. SC 1.4.13 states that : you can disclose the dropdown menu without moving the mouse. It also states that you can move the mouse onto the newly opened element. I have a doubt here : should we still be able to discard the element while the mouse pointer has already moved from the button to the additional content, without moving back to the button?
4. In a keyboard only scenario, if the dropdown menu automatically activates the submenu when the focus goes onto the toggle button, does 1.4.13 enforces the submenu to be dismissable without moving back to the same button after the keyboard focus has already been moved into the submenu?

In short, does a simple toggle button can satisfy this criteria without an Esc key? Also I perfectly understand that the Esc key can prevent a lot of useless interactions.

**Comment by mraccess77** [comment]
> so generally, the most accepted way to achieve that is to rely on keyboard Esc or similar. technically yes, clicking would also be a valid way to achieve that.

The click would have to be done in a way that doesn't move pointer hover - perhaps a click on the element without then causing something else to occur like a change of keyboard focus. So clicking is really only option when it's like a click to show and click to hide and not some other action.

In addition, I think folks are saying that if the additional content is shown on focus it could still be dismissed with click and pass 1.4.13 but it would not pass 2.1.1.  

Finally, I don't think the SC says if displayed on hover it can't be dismissed only on keyboard.
